server/util/grpc_client/grpc_client.go

```diff
@@ -216,3 +224,55 @@ func CommonGRPCClientOptions() []grpc.DialOption {
 		}),
 	}
 }
+
+type grpcClientConnPoolCache struct {
+	env         environment.Env
+	connMutex   *sync.Mutex
+	connPoolMap map[string]interfaces.ClosableClientConn
+}
+
+func RegisterConnPoolCache(env environment.Env) {
+	c := NewGrpcClientConnPoolCache(env)
+	env.SetGrpcClientConnPoolCache(c)
+}
+
+func isPermittedForDial(target string) bool {
+	u, err := url.Parse(target)
+	if err != nil {
+		return false
+	}
+	if cache_api_url.String() == "" {
+		return true
+	}
+
+	return u.Hostname() == "localhost" || (urlutil.GetDomain(u.Hostname()) == urlutil.GetDomain(cache_api_url.WithPath("").Hostname()))
+}
+
+func (cpc *grpcClientConnPoolCache) GetGrpcClientConnPoolForURL(target string) (conn interfaces.ClosableClientConn, err error) {
+	cpc.connMutex.Lock()
+	defer cpc.connMutex.Unlock()
+	connPool, ok := cpc.connPoolMap[target]
+	if ok && connPool != nil {
+		return connPool, nil
+	}
+	if !isPermittedForDial(target) {
+		return nil, status.InvalidArgumentErrorf("Tried to download from an unpermitted domain: %s", target)
+	}
+
+	// We didn't find a connection pool, so we'll make one.
+	connPool, err = DialInternal(cpc.env, target)
+	if err != nil {
+		return nil, err
+	}
+	cpc.connPoolMap[target] = connPool
+	log.Infof("Cached connection pool for target: %s, %d targets cached so far", target, len(cpc.connPoolMap))
+	return connPool, nil
+}
+
+func NewGrpcClientConnPoolCache(env environment.Env) *grpcClientConnPoolCache {
+	return &grpcClientConnPoolCache{
+		env:         env,
+		connMutex:   &sync.Mutex{},
```

Comment:
nit: it's more idiomatic to have mutexes be non-pointer types in structs, you wouldn't need this initializing code if that was the case

Reply:
Thanks, if there's one thing I'm not it's Idiomatic With Go